poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+
+// These are messages used for signatures
+var sigMsg []byte = []byte("Promise Signature")
+var sigBlameMsg []byte = []byte("Promise Blame Signature")
+
+/* Promise structs are mechanisms by which a server can promise other servers
+ * that an abstract.Secret will be availble even if the secret's owner goes
+ * down. The secret to be promised will be sharded into shared secrets that can
+ * be combined using Lagrange Interpolation to produce the original secret.
+ * Other servers will act as insurers maintaining a share. If a client ever
+ * needs the secret to be reconstructed, it can contact the insurers to regain
+ * the shares and reconstruct the secret.
+ *
+ * Note to Developers:
+ *       The insurers and secrets arrays should remain synchronized. In other
+ *       words, insurers[i] and secrets[i] should both refer to the same server.
+ *
+ * Note to users of this code:
+ *
+ *   Here is a list of methods that should be called by each type of server:
+ *
+ * - Promisers
+ *   * ConstructPromise
+ *
+ * - Insurers
+ *   * VerifyPromise
+ *   * VerifyShare
+ *   * Sign
+ *   * RevealShare
+ *   * Blame
+ *
+ * - Clients
+ *   * VerifyPromise
+ *   * VerifyRevealedShare
+ *
+ * - All
+ *   * UnmarshalInit
+ *   * VerifySignature
+ *   * VerifyBlame
+ */
+type Promise struct {
+
+	// The id of the promise used to differentiate it from others
+	// The id is the short term public key of the private key being promised
+	id string
+
+	// The cryptographic suite to use for the shared secrets
+	shareSuite abstract.Suite
+
+	// The suite used for the long term keys stored in the Promise
+	keySuite abstract.Suite
+
+	// The minimum number of shares needed to reconstruct the secret
+	t int
+
+	// The minimum number of PromiseSignatures approving the Promise that
+	// are needed before the Promise is certified. t <= r <= n
+	r int
+
+	// The total number of shares
+	n int
+
+	// The long-term public key of the promiser
+	pubKey abstract.Point
+
+	// The public polynomial that is used to verify the shared secrets
+	pubPoly poly.PubPoly
+
+	// A list of servers who will act as insurers of the Promise. The list
+	// contains the long-term public keys of the insurers
+	insurers []abstract.Point
+
+	// The list of shared secrets to be sent to the insurers. They are
+	// encrypted with Diffie-Hellman shared secrets between the insurer
+	// and the promiser.
+	secrets []abstract.Secret
+}
+
+/* Constructs a new Promise to guarentee a secret.
+ *
+ * Arguments
+ *    secretPair   = the keypair of the secret to be promised
+ *    longPair     = the long term keypair of the promiser
+ *    t            = minimum number of shares needed to reconstruct the secret.
+ *    r            = minimum PromiseSignatures needed to certify the Promise
+ *    insurers     = a list of the long-term public keys of the insurers.
+ *
+ * Returns
+ *   A newly constructed Promise
+ */
+func (p *Promise) ConstructPromise(secretPair *config.KeyPair,
```

Comment:
My initial thought on having ConstructPromise and UnmarshalInit be two different functions was to follow the model of sharing.go (an example is included at the bottom of the comment).

I was also hoping to make the code generalizable for future iterations. With the insurers list for example, it would be difficult for all nodes to know which insurers an individual promiser selected. This would likely differ from server to server since we would gain more security overall if each server had a different set of insurers. Hence, the nodes would need to wait to unmarshal the promise to figure out its insurers.

Concerning the longPair, an insurer would also have to remember that longPair should not be its own long term key pair but the pair from the server who owned the Promise they were trying to unmarshal.

Since what you would input for secretPair, longPair, and insurers would differ depending on whether you were a promiser taking out your own secret or an insurer unmarshalling a promise, I thought it might be easier for future users of this code base to have two different functions.

    // Initialize to an empty polynomial for a given group and threshold (degree),
    // typically before using Decode() to fill in from a received message.
    func (pub *PubPoly) Init(g abstract.Group, k int, b abstract.Point) {  
  	    pub.g = g
   	    pub.b = b
	    pub.p = make([]abstract.Point, k)
    }

    // Initialize to a public commitment to a given private polynomial.
    // Create commitments as encryptions of a given base point b,
    // or the standard base if b == nil.
    func (pub *PubPoly) Commit(pri *PriPoly, b abstract.Point) *PubPoly {
	    g := pri.g
	    k := len(pri.s)
	    p := make([]abstract.Point, k)
	    for i := 0; i < k; i++ {
		    p[i] = g.Point().Mul(b, pri.s[i])
	    }
   	    pub.g = g
	    pub.b = b
   	    pub.p = p
	    return pub
    }